### ISSUE TITLE
Unexpected failure on data schema validation (GN)

### ISSUE THREAD

**Issue by PeterParslow** [inspire-eu-validation/ets-repository#121]
I have a test run report at http://inspire-sandbox.jrc.ec.europa.eu/etf-webapp/v2/TestRuns/EIDecb7974c-0b4a-4949-a9a0-4ee8bae9f76f.html, which gives a failure at gmlas.b.2:

The dataset has 1 file(s) with errors for this assertion.
XML document 'HP40.gml': The file has 2 schema validation error(s).
XML document 'HP40.gml': -1:-1: Premature end of file.
XML document 'HP40.gml': -1:-1: Premature end of file.

The sample file which I uploaded contains this schemaLocation statements:

xsi:schemaLocation="http://www.opengis.net/gml/3.2 http://schemas.opengis.net/gml/3.2.1/gml.xsd http://namespaces.os.uk/open/names/1.0/ http://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd" 

Both those files seem accesible & valid, so what is the "premature end of file"?

**Comment by jonherrmann** [comment]
I'm not sure if the schema validator has a problem with the http to https redirect of the second schema.

Could you mail the file to etf@interactive-instruments.de ?

**Comment by jonherrmann** [comment]
Thanks for the file, Peter. 

I changed the URL to https and now the schema validator validates the file without the "Premature end of file" error. So we have an issue in the validator with the URL redirect handling. We are currently using the integrated schema validator of the XML database but may need to improve the validator engine (see the first draft in issue https://github.com/interactive-instruments/etf-webapp/issues/85 ).

Regardless, there seems to be an issue with the schema :
http://inspire-sandbox.jrc.ec.europa.eu/etf-webapp/v2/TestRuns/EID3219bfb4-c76e-40ae-b6a5-f38641910c73.html?lang=en#EIDb7871996-8a71-4927-ba29-acc9e9619b95

**Comment by PeterParslow** [comment]
I guess I took too long to get to this, because the link you provide now has no contents.

However, I re-ran the test, on a small sample of the data fixed to avoid redirection, and got an error against the schema (http://inspire-sandbox.jrc.ec.europa.eu/etf-webapp/v2/TestRuns/EID9d5c8f15-00f5-49f3-940d-a3db5b94ec29.html). I can't work out what the actual error is: the most likely candidate, which then cascades to the others is "Failed to read schema document 'https://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd', because 1) could not find the document; 2) the document could not be read; 3) the root element of the document is not <xsd:schema>."

Which is odd, because I can browse to the file, and the root is <schema> - the implicit XSD namespace is no less present than it is in INSPIRE's or OGC's own XSDs (for example). I haven't met a validator which requires XSDs to explicitly define XSD itself, which may be what this message is complaining about - or perhaps it just can't load the file at all?

**Comment by cportele** [comment]
I downloaded the OS Open Names dataset and tried to validate it locally using Xerces-J.

Without the http->https edit I get (as expected):

> *** Fatal Error: org.xml.sax.SAXParseException; systemId: http://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd; lineNumber: 1; columnNumber: 1; Premature end of file.
> (http://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd, line 1, column 1)
> Parsing error: org.xml.sax.SAXParseException; systemId: http://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd; lineNumber: 1; columnNumber: 1; Premature end of file.

With the edit the validation passes. 

So either the OSOpenNames.xsd file cannot be accessed from the sandbox or it is an issue with the built-in XML validator in Java (which is used). I will check, which of the two is the case.

**Comment by cportele** [comment]
The sandbox cannot access https://www.ordnancesurvey.co.uk/xml/open/names/1.0/OSOpenNames.xsd.

I have made the XSD available at http://portele.de/OSOpenNames.xsd and changed HP40.gml to reference this schema document. Then the validation passes:

http://inspire-sandbox.jrc.ec.europa.eu/etf-webapp//v2/TestRuns/EIDedb2d45e-8c89-45e3-a40c-21fa3335f0b9.html?lang=en

I.e., this must be another certificate issue.

**Comment by PeterParslow** [comment]
Thanks for the explanation. If the certificate issue is at our end, I can look into that - but I can inspect the certificate (with Firefox built in tools), and it looks OK to my untrained eyes.

**Comment by cportele** [comment]
@PeterParslow My guess is that it is a similar issue to #86, #89, #128, #131.

@robsgnao

**Comment by PeterParslow** [comment]
That does look likely, and suggests that the ETS needs a mechanism (& authority?) for adding Certificate Authorities?

**Comment by robsgnao** [comment]
@cportele GeoTrust intermediate certificate was missing (Extended Validation SHA256), just added; validation should be working fine now.

**Comment by cportele** [comment]
@robsgnao Yes, validation is now working (when using the https-URL of the XSD).

**Comment by robsgnao** [comment]
The SSL-related issue is not present while using OpenJDK instead of Oracle JDK. Validator hosted at JRC has been switched to OpenJDK yesterday.
For the time being, this change seems to be sufficient, and a mechanism for adding & trusting Root and Intermediate certificates of Certification Autorities seems to not be needed.
We will monitor this issue over the coming week to verify this assumption.
